WIP: bindings: add python3 bindings for libelektra-tools

manuelm opened the pull request:
this is an early draft to play with the libelektra tools bindings for python3.
e.g. this works:
```
PYTHONPATH="." /usr/bin/python3 -c "import tools; import kdb; b=tools.Backend(); b.setMountpoint(kdb.Key('/', kdb.KEY_CASCADING_NAME, kdb.KEY_END), kdb.KeySet(0)); b.addPlugin('resolver'); b.addPlugin('dump'); b.validated(); x=kdb.KeySet(); b.serialize(x); [ print(y) for y in x ]"
system/elektra/mountpoints/\/
system/elektra/mountpoints/\//config
system/elektra/mountpoints/\//config/path
system/elektra/mountpoints/\//errorplugins
system/elektra/mountpoints/\//errorplugins/#5#resolver#resolver#
system/elektra/mountpoints/\//getplugins
system/elektra/mountpoints/\//getplugins/#0#resolver
system/elektra/mountpoints/\//getplugins/#5#dump#dump#
system/elektra/mountpoints/\//mountpoint
system/elektra/mountpoints/\//setplugins
system/elektra/mountpoints/\//setplugins/#0#resolver
system/elektra/mountpoints/\//setplugins/#5#dump
system/elektra/mountpoints/\//setplugins/#7#resolver
```
... But I have no idea if it's correct :-)

Things to do:
- [ ] rename python3 to python before (in other PR)
- [ ] make tools a submodule of kdb so it's kdb.tools
  * create a directory `kdb`
  * rename `kdb.py` to `kdb/__init__p.y`
  * move `_kdb.so` to `kdb/_kdb.so`
  * move `tools.py` + `_tools.so` to `kdb/`
 - [ ] fix binding of kdb::tools::Modules::load ?
 - [ ] also bind helper + merging stuff ?

manuelm commented:
rebased as well

markus2330 commented:
jenkins build bindings please

markus2330 commented:
@manuelm can we bring this to a mergable state?

Note that the 3-way merging is now in a separate lib.

manuelm commented:
I can fix the conflicts and make it compile+work again. But I won't put any more development into this. Fine for you?

markus2330 commented:
Of course, if mounting works it is already much more than nothing.

markus2330 commented:
@tom-wa will then use this in the Ansible binding.

manuelm commented:
As promised I have reworked the python bindings for libelektratools. As the C++ code got rewritten I have pretty much started over. However with the help of the current ruby bindings things went quite smoothly.

This time I have also added tests. Examples can be added by another person. Or simply use the documentation of the C++ interface.

mpranj commented:
Thank you for working on this! It seems that it does not compile. Can you please look into that?

At first glance it looks good to me. I will review once the builds run through.

manuelm commented:
The latest commit should fix the compile issue. I'm just faster than our CI. Actually any human being is faster.

mpranj commented:
jenkins build libelektra please

mpranj commented:
Cannot confirm. Formatting check fails.

manuelm commented:
This one: https://github.com/ElektraInitiative/libelektra/pull/323/checks?check_run_id=347373113 ? How do I tell Cirrus to ignore the generated code?

mpranj commented:
Yes that one. I think cirrus will only compile what is declared to be compiled by cmake.

Also there is a formatting error in the CMakeLists https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-323/5/pipeline/353 .

manuelm commented:
the suggested formatting for the comment is kind of weird

mpranj commented:
> the suggested formatting for the comment is kind of weird

You're right. cmake-format wants you to surround the lists with empty lines.

mpranj commented on src/bindings/swig/python/CMakeLists.txt:
```diff
@@ -35,15 +35,31 @@ else ()
+
+		# TODO: Reenable the following warning after we add a virtual destructor to `PluginDatabase`, and its subclasses.
+		# See also:
+		# - https://github.com/ElektraInitiative/libelektra/pull/1841
```
```suggestion
		#
		# - https://github.com/ElektraInitiative/libelektra/pull/1841
```

mpranj commented on src/bindings/swig/python/CMakeLists.txt:
```diff
@@ -35,15 +35,31 @@ else ()
+		# TODO: Reenable the following warning after we add a virtual destructor to `PluginDatabase`, and its subclasses.
+		# See also:
+		# - https://github.com/ElektraInitiative/libelektra/pull/1841
+		# - https://github.com/ElektraInitiative/libelektra/commit/15d67328
```
```suggestion
		# - https://github.com/ElektraInitiative/libelektra/commit/15d67328
		#
```

manuelm commented:
Thanks. Lets see what's wrongly formatted now :-)

mpranj commented on src/bindings/swig/python/CMakeLists.txt:
```diff
@@ -45,18 +45,19 @@ else ()
+	# TODO: Reenable the following warning after we add a virtual destructor to `PluginDatabase`, and its subclasses.
+	# See also:
+	#
+	# * https://github.com/ElektraInitiative/libelektra/pull/1841
```
Actually the "-" bullet points were correct according to the cmake format file, but let's see what it says now.

manuelm replied:
hahaha

manuelm commented:
jenkins build libelektra please

manuelm commented:
@mpranj Finally the CI should be green in a few minutes. One thing you might want to look at: No Jenkins setup/build provides SWIG 4. So Jenkins never builds the new bindings. However Travis does.

mpranj commented:
Thank you for the update!

This essentially means we have no python bindings for debian any more. 😢Was this discussed with markus already?

I would like to add a jenkins build with fedora 31. Do you mind if we at least first do that, which would allow us to use fedora + SWIG 4 to test the bindings on our infrastructure too? Ideally we would not drop python bindings on debian, but I don't know what can be done about that.

manuelm commented:
> This essentially means we have no python bindings for debian any more. cryWas this discussed with markus already?

No, don't worry. Only the tools bindings require SWIG 4.

mpranj commented:
Why does the debian package build complain with this then:

```
dh_install --list-missing
dh_install: Please use dh_missing --list-missing/--fail-missing instead
dh_install: This feature will be removed in compat 12.
dh_install: Cannot find (any matches for) "usr/lib/python3/*-packages/_kdb.so" (tried in ., debian/tmp)

dh_install: python3-elektra missing files: usr/lib/python3/*-packages/_kdb.so
dh_install: Cannot find (any matches for) "usr/lib/python3/*-packages/kdb.py" (tried in ., debian/tmp)

dh_install: python3-elektra missing files: usr/lib/python3/*-packages/kdb.py
dh_install: missing files, aborting
```

It looks like the kdb is missing, or did some paths change here?

manuelm commented:
Yeah. kdb is a submodule now. So it's `/path/to/site-packages/kdb/{kdb.so,kdb.py,tools.py,tools.so,__init__.py}`

markus2330 commented:
Great work!

@tom-wa can you check if it works for you?

markus2330 commented:
@mpranj Please adapt the Debian files, so that (the non-tool parts) are still part of the Debian package. Unfortunately, there is no smooth migration path for that, so at some point master will not build. (When the debian folder is changed and this PR is not yet merged).

mpranj commented:
> Please adapt the Debian files ...

Done. The build went through now.

markus2330 commented:
I think for @tom-wa it will be easier to test if it is already merged (and ideally packages are available, I hope @Mistreated soon has progress on that).

mpranj commented:
@manuelm thank you so much for finishing this! :tada:
